external-crates/move/crates/move-compiler/src/parser/syntax.rs

```diff
@@ -274,45 +349,134 @@ fn parse_comma_list_after_start<F, R>(
+                        context,
+                        start_token,
+                        end_token,
+                        /* separator */ Tok::Comma,
+                        /* for list */ true,
+                        *diag,
+                    );
+                }
+            }
+        } else {
             let current_loc = context.tokens.start_loc();
             let loc = make_loc(context.tokens.file_hash(), current_loc, current_loc);
-            return Err(Box::new(diag!(
+            let diag = diag!(
                 Syntax::UnexpectedToken,
                 (loc, format!("Expected {}", item_description))
-            )));
+            );
+            separated_items_error_advance(
+                context,
+                start_token,
+                end_token,
+                /* separator */ Tok::Comma,
+                /* for list */ true,
+                diag,
+            );
         }
-        v.push(parse_list_item(context)?);
-        adjust_token(context.tokens, end_token);
-        if match_token(context.tokens, end_token)? {
-            break Ok(v);
+        // The stop set check is done at the end of the loop on purpose as we need to attempt
+        // parsing before checking it. If we do not, in the best case, we will get a less meaningful
+        // error message if the item belongs to the token set incorrectly (e.g., `fun` keyword), and
+        // in the worst case, we will get an error in the correct code (e.g., if a function argument
+        // is named `entry`)
+        if context.at_stop_set() {
+            break;
         }
-        if !match_token(context.tokens, Tok::Comma)? {
-            let current_loc = context.tokens.start_loc();
-            let loc = make_loc(context.tokens.file_hash(), current_loc, current_loc);
-            let loc2 = make_loc(context.tokens.file_hash(), start_loc, start_loc);
-            return Err(Box::new(diag!(
-                Syntax::UnexpectedToken,
-                (loc, format!("Expected '{}'", end_token)),
-                (loc2, format!("To match this '{}'", start_token)),
-            )));
+    }
+    if consume_token(context.tokens, end_token).is_err() {
+        let current_loc = context.tokens.start_loc();
+        let loc = make_loc(context.tokens.file_hash(), current_loc, current_loc);
+        let loc2 = make_loc(context.tokens.file_hash(), start_loc, start_loc);
+        context.add_diag(diag!(
+            Syntax::UnexpectedToken,
+            (loc, format!("Expected '{}'", end_token)),
+            (loc2, format!("To match this '{}'", start_token)),
+        ));
+    }
+    v
+}
+
+/// Attempts to skip tokens until the end of the item in a series of separated (which started with
+/// an already consumed starting token) - looks for a matched ending token or a token appearing
+/// after the separator. This helper function is used when parsing lists and sequences.
+fn separated_items_error_advance(
```

Comment:
Small nit, but most places in this code start with the action as the first word.
```suggestion
fn advance_separated_items_error(
```